### PR TITLE
refactor: Telegram/tglib for Hono

### PR DESCRIPTION
**Pull request by synchrone**
## Involved Issue / 该 PR 相关 Issue

Followup #13255 after refactoring for TypeScript and Hono

## Example for the Proposed Route(s) / 路由地址示例

```routes
/telegram/channel/telegram
```

## 新 RSS 路由检查表 / New RSS Route Checklist
  
- [X] 新的路由 New Route
  - [X] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [X] 文档说明 Documentation
  - [X] 中文文档 CN
  - [X] 英文文档 EN
- [ ] 全文获取 fulltext
  - [ ] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [X] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [X] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note

Passing app in route registration to get lower level access for streaming big media which needs HTTP Range requests

### CODE REVIEW

File: lib/routes/telegram/tglib/client.ts [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+import { fileURLToPath } from 'node:url';
+import { Api, TelegramClient } from 'telegram';
+import { StringSession } from 'telegram/sessions';
 
 const apiId = Number(config.telegram.apiId ?? 4);
 const apiHash = config.telegram.apiHash ?? '014b35b6184100b085b0d0572f9b5103';
 
 const stringSession = new StringSession(config.telegram.session);
-const client = new TelegramClient(stringSession, apiId, apiHash, {
+export const client = new TelegramClient(stringSession, apiId, apiHash, {
```
Comment (TonyRL) [inline lib/routes/telegram/tglib/client.ts:12]: As mentioned in https://github.com/DIYgod/RSSHub/pull/13255#discussion_r1409327692, `telegram` shouldn't start with RSSHub.

You can achieve the similar thing like:
```ts
export let client: TelegramClient;

export const login = () => {
    if (!client && config.telegram.session) {
        client = new TelegramClient(..., {... });
        client.start(...);
    }
}
```
and call `login()` in your `handler`.

File: lib/routes/telegram/tglib/client.ts [hunk 2]
```diff
@@ -24,73 +24,7 @@ if (config.telegram.session) {
     });
```
Comment (TonyRL) [inline lib/routes/telegram/tglib/client.ts:24]: https://github.com/DIYgod/RSSHub/blob/94a78a70bf72463460c3ff17d7b2872f9236981f/lib/routes/telegram/tglib/client.ts#L22

Do not throw plain string, throw `new Error(contain)` instead.

File: lib/routes/telegram/tglib/channel.ts [hunk 3]
```diff
@@ -160,6 +82,4 @@ export default async (ctx) => {
         allowEmpty: ctx.req.param('id') === 'allow_empty',
         description: `@${channelInfo.username} on Telegram`,
     });
-};
-
-module.exports.getMedia = getMedia;
+}
```
Comment (TonyRL) [inline lib/routes/telegram/tglib/channel.ts:85]: https://github.com/DIYgod/RSSHub/blob/94a78a70bf72463460c3ff17d7b2872f9236981f/lib/routes/telegram/tglib/channel.ts#L65

Do not slice `title`, RSSHub will handle it.
https://github.com/DIYgod/RSSHub/blob/7a8335ee1e12187e483b5685756cce3fb37d42bf/lib/middleware/template.ts#L41-L47

File: lib/routes/telegram/tglib/channel-media.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        ctx.header('Content-Type', doc.mimeType);
+        ctx.header('Accept-Ranges', 'bytes');
+
+        const rangeHeader = ctx.req.header('Range');
+        const range = parseRange(rangeHeader, doc.size.valueOf() - 1);
+        if (range.length > 1) {
+            return ctx.text('Not Satisfiable', 416);
+        }
+
+        let stream;
```
Comment (TonyRL) [inline lib/routes/telegram/tglib/channel-media.ts:206]: Please add CSP header.

**Comment by synchrone** [comment]
@TonyRL sorry i got carried away with all the TS changes and accidentally fixed some niche use-cases which ended up as new features.
Should I split this into multiple PRs ?

File: lib/routes/telegram/tglib/channel-media.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+    const doc = getDocument(media);
+    if (doc) {
+        if ('thumb' in ctx.req.query()) {
+            ctx.header('Content-Type', 'image/jpeg');
+            return streamResponse(ctx, streamThumbnail(client, doc));
+        }
+        ctx.header('Content-Type', doc.mimeType);
+        ctx.header('Accept-Ranges', 'bytes');
+        ctx.header('Content-Security-Policy', "default-src 'self'");
```
Comment (TonyRL) [inline lib/routes/telegram/tglib/channel-media.ts:197]:
```suggestion
        ctx.header('Content-Security-Policy', "default-src 'self'; script-src 'none'");
```

**Comment by synchrone** [comment]
TODO: find a new way to register hono route instead of autodiscovery in `lib/registry.ts`

**Comment by synchrone** [comment]
@TonyRL I see my implementation is now in actual use. 
If you will commit the time to review this PR I can update if from master. There's good new functionality here, and, importantly, an optimized video streaming.

**Pull request closed without merging**